joint_limits/test/test_simple_joint_limiter.cpp

```diff
@@ -12,31 +12,266 @@
+    ASSERT_TRUE(joint_limiter_->enforce(current_joint_states_, desired_joint_states_, period));
+
+    // check if pos limits applied
+    CHECK_STATE_SINGLE_JOINT(
+      desired_joint_states_, 0,
+      5.0,                                    // pos unchanged
+      desired_joint_states_.velocities[0],    // vel unchanged
+      desired_joint_states_.accelerations[0]  // acc unchanged
+    );
+  }
+}
+
+TEST_F(SimpleJointLimiterTest, when_position_close_to_pos_limit_expect_deceleration_enforced)
+{
+  SetupNode("simple_joint_limiter");
+  Load();
+
+  if (joint_limiter_)
+  {
+    Init();
+    Configure();
+
+    // using 0.05 because 1.0 sec invalidates the "small dt integration"
+    rclcpp::Duration period(0, 50000000);  // 0.05 second
+
+    // close to pos limit should reduce velocity and stop
+    current_joint_states_.positions[0] = 4.851;
+    current_joint_states_.velocities[0] = 1.5;
+    desired_joint_states_.positions[0] = 4.926;
+    desired_joint_states_.velocities[0] = 1.5;
+
+    // this setup requires 0.15 distance to stop, and 0.2 seconds (so 4 cycles at 0.05)
+    for (auto i = 0u; i < 4; ++i)
+    {
+      auto previous_vel_request = desired_joint_states_.velocities[0];
+      ASSERT_TRUE(joint_limiter_->enforce(current_joint_states_, desired_joint_states_, period));
+
+      ASSERT_LE(
+        desired_joint_states_.velocities[0],
+        previous_vel_request);  // vel adapted to reach end-stop should be decreasing
+      // NOTE: after the first cycle, vel is reduced and does not trigger stopping position limit
+      // hence no max deceleration anymore...
+      ASSERT_LE(
+        desired_joint_states_.positions[0],
+        5.0 +
+          10.0 *
+            COMMON_THRESHOLD);  // is not decelerating at max all the time, hence err in last cycle
+      ASSERT_LE(desired_joint_states_.accelerations[0], 0.0);  // should decelerate
+
+      Integrate(period.seconds());
+
+      ASSERT_LE(current_joint_states_.positions[0], 5.0);  // below joint limit
+    }
+  }
+}
+
+TEST_F(SimpleJointLimiterTest, when_acceleration_exceeded_expect_acc_enforced)
+{
+  SetupNode("simple_joint_limiter");
+  Load();
+
+  if (joint_limiter_)
+  {
+    Init();
+    Configure();
+
+    rclcpp::Duration period(0, 50000000);
+
+    // desired acceleration exceeds
+    current_joint_states_.positions[0] = 0.0;
+    current_joint_states_.velocities[0] = 0.0;
+    desired_joint_states_.velocities[0] = 0.5;  // leads to acc > max acc
+    ASSERT_TRUE(joint_limiter_->enforce(current_joint_states_, desired_joint_states_, period));
+
+    // check if vel and acc limits applied
+    CHECK_STATE_SINGLE_JOINT(
+      desired_joint_states_, 0,
+      desired_joint_states_.positions[0],  // pos unchanged
+      0.25,                                // vel limited max acc * dt
+      5.0                                  // acc max acc
+    );
+  }
+}
+
+TEST_F(SimpleJointLimiterTest, when_deceleration_exceeded_expect_acc_enforced)
```

Comment:
I don't understand the description - if deceleration is exceeded then we would expect deceleration limits to work. If deceleration is not defined then acceleration limits should be used.

Reply:
Oh, I think there is only a small issue in naming. Is it?

Reply:
typical copy paste typo